sky/shell/platform/ios/semantics_listener_impl.mm

```diff
@@ -0,0 +1,245 @@
+    id = node->id;
+  }
+  assert(id == node->id);
+
+  if (!node->flags.is_null()) {
+    canBeTapped = node->flags->canBeTapped;
+    canBeLongPressed = node->flags->canBeLongPressed;
+    canBeScrolledHorizontally = node->flags->canBeScrolledHorizontally;
+    canBeScrolledVertically = node->flags->canBeScrolledVertically;
+    hasCheckedState = node->flags->hasCheckedState;
+    isChecked = node->flags->isChecked;
+  }
+
+  if (!node->strings.is_null()) {
+    if (label) {
+      delete label;
+    }
+    label = (char*)malloc((node->strings->label.size() + 1) * sizeof(char));
+    std::strcpy(label, node->strings->label.data());
+  }
+
+  if (!node->geometry.is_null()) {
+    if (!node->geometry->transform.is_null()) {
+      float* values = asFloatArray(node->geometry->transform.Pass());
+      transform.setColMajorf(values);
+      delete values;
+    }
+    left = node->geometry->left;
+    top = node->geometry->top;
+    width = node->geometry->width;
+    height = node->geometry->height;
+  }
+
+  if (!node->children.is_null()) {
+    const std::vector<AccessibilityNode*>& oldChildren = children;
+    children = std::vector<AccessibilityNode*>();
+    for (AccessibilityNode* child : oldChildren) {
+      assert(child->parent != NULL);
+      child->parent = NULL;
+      // TODO(tvolkert): recursively set parent to null on descendant
+      // tree, or assert may fail in removePersistentNode()
+    }
+    for (const semantics::SemanticsNodePtr& childNode : node->children) {
+      AccessibilityNode* child = outer->updateSemanticsNode(childNode);
+      child->parent = this;
+      children.push_back(child);
+    }
+    for (AccessibilityNode* child : oldChildren) {
+      if (child->parent == NULL) {
+        outer->removePersistentNode(child);
+      }
+    }
+  }
+
+  // Has to be done after children are updated
+  // since they also get marked dirty
+  invalidateGlobalGeometry();
+}
+
+float* asFloatArray(mojo::Array<float> input) {
+  float* result = (float*)malloc(input.size() * sizeof(float));
+  for (size_t i = 0; i < input.size(); i++) {
+    result[i] = input[i];
+  }
+  return result;
+}
+
+void SemanticsListenerImpl::AccessibilityNode::invalidateGlobalGeometry() {
+  if (geometryDirty) {
+    return;
+  }
+  geometryDirty = true;
+  for (AccessibilityNode* child : children) {
+    child->invalidateGlobalGeometry();
+  }
+}
+
+SkMatrix44& SemanticsListenerImpl::AccessibilityNode::getGlobalTransform() {
+  if (geometryDirty) {
+    if (parent == NULL) {
+      globalTransform = transform;
+    } else {
+      SkMatrix44 parentTransform = parent->getGlobalTransform();
+      if (transform.isIdentity()) {
+        globalTransform = parentTransform;
+      } else if (parentTransform.isIdentity()) {
+        globalTransform = transform;
+      } else {
+        globalTransform = transform * parentTransform;
+      }
+    }
+  }
+  return globalTransform;
+}
+
+CGRect SemanticsListenerImpl::AccessibilityNode::getGlobalRect() {
```

Comment:
There must be helper functions for this which can avoid this.  Imagine you typed 0 instead of 1, thsi would be very hard to debug. :(